File: Python_Solutions/Average_of_Levels_in_Binary_Tree.py
"""
Leetcode Challenge: Average of Levels in Binary Tree

Given the root of a binary tree, return the average value of the nodes on each level 
in the form of an array. Answers within 10^-5 of the actual answer will be accepted.

6/16/24
Mark Robuck
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    """
    Average the node values of each level of a binary tree.
    Args:
        root(TreeNode): given tree to find averages for
    Returns:
        List(float): average of each level as a list
    """
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        averages = []
        queue = [root]
        #if the tree is empty
        if not root:
            return averages
        
        while queue:
            level_sum = 0
            level_count = len(queue)
            next_queue = []
            #sum single level in the tree
            for node in queue:
                level_sum += node.val
                #add next level of nodes to the next queue
                if node.left:
                    next_queue.append(node.left)
                if node.right:
                    next_queue.append(node.right)
            #calculate average for the level and append to the list
            averages.append(level_sum / level_count)
            #move to the next level of nodes
            queue = next_queue
        
        return averages

File: Java_Solutions/Average_of_Levels_in_Binary_Tree.java
/*
* Leetcode Challenge: Average of Levels in Binary Tree
*
* Given the root of a binary tree, return the average value of the nodes on each level 
* in the form of an array. Answers within 10^-5 of the actual answer will be accepted.
*
* 6/16/24
* Mark Robuck
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
    * Average the node values of each level of a binary tree.
    * @params root, given tree to find averages for
    * @ return average of each level as a list
    */
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> averages = new ArrayList<>();
        //if tree is empty
        if (root == null) {
            return averages;
        }
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            double levelSum = 0;
            //sum single level in the tree
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                levelSum += node.val;
                //add next level of nodes to the next queue
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
            //calculate average for the level and append to the list
            averages.add(levelSum / levelSize);
            }
        
        return averages;
    }
}
